{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Configurable Variables\n",
    "Variables that can be modified in this notebook\n",
    "- `input_data` — path to the input folder containing the dataset\n",
    "- `supress_warnings` — boolean flag to enable/disable warning messages (`True`/`False`)\n",
    "---\n",
    "**Set following variables if you are running notebook interactively.**"
   ],
   "id": "75dd8452d71f2682"
  },
  {
   "metadata": {
    "collapsed": true
   },
   "cell_type": "code",
   "source": [
    "input_path_folder_default = \"output_datasets/SN048_A121573_Rep1/\"\n",
    "suppress_warnings_default = False\n",
    "grid_size_default = 21"
   ],
   "id": "26f66cb190286669",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# SPATIAL TRANSCRIPTOMICS – IDENTIFICATION OF SVGs WITH SOMDE\n",
    "\n",
    "Spatial transcriptomics links gene expression to tissue location. To find **spatially variable genes (SVGs)**, we use **SOMDE**, a fast method that detects spatial expression patterns using **self-organizing maps** and **statistical modeling**. This notebook demonstrates how to load data, run SOMDE, and visualize the resulting SVGs.\n"
   ],
   "id": "76384a3ea5e69110"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Load variables from CMD",
   "id": "8704f8d5a27f0ecd"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "SUPPRESS_WARNINGS = None\n",
    "try:\n",
    "    if suppress_warnings == \"True\":\n",
    "        SUPPRESS_WARNINGS = True\n",
    "    else:\n",
    "        SUPPRESS_WARNINGS = False\n",
    "except NameError:\n",
    "    SUPPRESS_WARNINGS = suppress_warnings_default\n",
    "\n",
    "try:\n",
    "    GRID_SIZE = grid_size\n",
    "except NameError:\n",
    "    GRID_SIZE = grid_size_default"
   ],
   "id": "4b29d7d23801d942",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Suppress Warnings (Optional)\n",
    "\n",
    "* Suppress warnings for cleaner output.\n",
    "* Comment out the line below if warnings are desired for debugging."
   ],
   "id": "684ddd57fe708fa1"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Suppress warnings for cleaner output\n",
    "import warnings\n",
    "import os\n",
    "\n",
    "if SUPPRESS_WARNINGS:\n",
    "    print(\"Ignoring warnings.\")\n",
    "    warnings.filterwarnings('ignore')"
   ],
   "id": "83880dbff772b24b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Environment setup\n",
    "\n",
    "1. Install the necessary packages using `pip` if they are not already installed:"
   ],
   "id": "ebafd741c78d4fbd"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "# use conda env from readme",
   "id": "80cb40bfdf4967b6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "2. Import necessary libraries:",
   "id": "b7750ab659a5ea6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# work with files libraries\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "\n",
    "# SOMDE libraries\n",
    "import somoclu\n",
    "from somde import plotgene\n",
    "from somde import SomNode\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy as sp\n",
    "import math\n",
    "\n",
    "# cpu usage and time statistics\n",
    "import psutil\n",
    "import time\n",
    "import platform"
   ],
   "id": "d7dc7e84196cb2ed",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "3. load the dataset\n",
    "  * **Option 1**: Import data directly from Google Drive (requires Google Colab).\n",
    "  * **Option 2**: Load dataset path from CMD."
   ],
   "id": "58ecc56ff525bc57"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# OPTION 1\n",
    "# from google.colab import drive\n",
    "# drive.mount('/content/drive')"
   ],
   "id": "9f29ea605dd10741",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# OPTION 2\n",
    "DATA_PATH = None\n",
    "try:\n",
    "    DATA_PATH = input_data_folder\n",
    "except NameError:\n",
    "    DATA_PATH = input_path_folder_default\n",
    "\n",
    "if DATA_PATH is None or not os.path.exists(DATA_PATH):\n",
    "    print(\"ERROR: Datapath (\", DATA_PATH ,\") not found.\")\n",
    "    sys.exit()\n",
    "\n",
    "print(\"Using input path:\", DATA_PATH)\n",
    "\n",
    "DATA_FOLDER_NAME = DATA_PATH.split(\"/\")[-2].strip()\n",
    "output_dir = os.path.join(\"somde_results/\" + DATA_FOLDER_NAME)\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "os.makedirs(\"figures/show/somde/\" + DATA_FOLDER_NAME, exist_ok=True)"
   ],
   "id": "8f09c7dd3ac3539d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "4. Get platform information",
   "id": "8f06b61807d183f3"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ---- System info ----\n",
    "cpu_model = platform.processor()\n",
    "cpu_cores = psutil.cpu_count(logical=False)\n",
    "cpu_threads = psutil.cpu_count(logical=True)\n",
    "\n",
    "total_ram = psutil.virtual_memory().total / (1024**3)  # GB\n",
    "\n",
    "print(\"CPU model:\", cpu_model)\n",
    "print(\"CPU cores:\", cpu_cores)\n",
    "print(\"CPU threads:\", cpu_threads)\n",
    "print(\"Total RAM: {:.2f} GB\".format(total_ram))\n",
    "\n",
    "# ---------------- SAVE TO CSV -----------------\n",
    "csv_file = os.path.join(output_dir, \"system_info.csv\")\n",
    "\n",
    "with open(csv_file, \"w\", newline=\"\") as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow([\"CPU_model\", \"CPU_cores\", \"CPU_threads\", \"Total_RAM_GB\"])\n",
    "    writer.writerow([cpu_model, cpu_cores, cpu_threads, round(total_ram, 2)])\n",
    "\n",
    "print(f\"System info saved to {csv_file}\")"
   ],
   "id": "79e90735390a1cb7",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# SOMDE analaysis",
   "id": "b97e627156f8b36e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Load gene expression data\n",
    "**Expecxted files**\n",
    "- *Expression counts (`count.csv`)* with spots as rows and genes as columns.\n",
    "- *Metadata (`idx.csv`)* containing spatial coordinates\n"
   ],
   "id": "d74d9af6f522d36d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Load expression counts\n",
    "df = pd.read_csv(os.path.join(DATA_PATH, \"count.csv\"), sep=',', index_col=0)\n",
    "print(\"Counts shape:\", df.shape)\n",
    "\n",
    "# Load metadata / spot info\n",
    "corinfo = pd.read_csv(os.path.join(DATA_PATH, \"idx.csv\"), sep=',', index_col=0)\n",
    "\n",
    "# Add total counts per spot (sum across genes) with proper alignment\n",
    "total_counts = df.sum(axis=1)\n",
    "corinfo[\"total_count\"] = total_counts.reindex(corinfo.index)\n",
    "print(corinfo.head())\n",
    "\n",
    "# Extract spatial coordinates\n",
    "X = corinfo[['x', 'y']].values.astype(np.float32)\n",
    "\n",
    "# Preview\n",
    "print(\"Counts matrix shape:\", df.shape)\n",
    "print(df.head())\n",
    "print(\"Spatial coordinates shape:\", X.shape)\n",
    "print(\"Metadata preview:\")\n",
    "print(corinfo.sort_values(by=['x', 'y'], ascending=True).head())\n",
    "\n",
    "num_samples = X.shape[0]\n",
    "print(\"Number of spots / samples:\", num_samples)"
   ],
   "id": "46b7ca4b8f0847a4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Initialize SOM Node\n",
    "\n",
    "This cell creates and trains a **Self-Organizing Map (SOM) node** using the spatial coordinate matrix `X`."
   ],
   "id": "b7f1bab381341c8a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "print(\"SOM node creation\")\n",
    "\n",
    "process = psutil.Process(os.getpid())\n",
    "process.cpu_percent(interval=None)   # reset counters\n",
    "\n",
    "start_cpu = time.process_time()\n",
    "start_wall = time.time()\n",
    "\n",
    "k = num_samples / (GRID_SIZE * GRID_SIZE)\n",
    "# -------- code to measure -----------\n",
    "som = SomNode(X, k)\n",
    "print(50 * \"-\")\n",
    "# ------------------------------------\n",
    "\n",
    "end_cpu = time.process_time()\n",
    "end_wall = time.time()\n",
    "\n",
    "cpu_time = end_cpu - start_cpu\n",
    "wall_time = end_wall - start_wall\n",
    "cpu_usage = process.cpu_percent(interval=None)\n",
    "ram_usage = process.memory_info().rss / (1024**2)  # MB\n",
    "\n",
    "print(\"CPU time: {:.6f} seconds\".format(cpu_time))\n",
    "print(\"Wall time: {:.6f} seconds\".format(wall_time))\n",
    "print(\"CPU usage (%):\", cpu_usage)\n",
    "print(\"RAM used by process: {:.2f} MB\".format(ram_usage))\n",
    "\n",
    "# ---------------- SAVE TO CSV -----------------\n",
    "with open(os.path.join(output_dir,  \"requirements.csv\"), \"w\", newline=\"\") as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow([\"Task\", \"CPU_time\", \"Wall_time\", \"CPU_usage_percent\", \"RAM_used_MB\"])\n",
    "    writer.writerow([\"somde node creation\", cpu_time, wall_time, cpu_usage, ram_usage])"
   ],
   "id": "49fc1f5080c88799",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Visualize Initial SOM Codebook\n",
    "\n",
    "This cell visualizes the **initial codebook** of the Self-Organizing Map (SOM)."
   ],
   "id": "fc99939019a3b563"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "som.viewIniCodebook()\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(5, 5)\n",
    "fig.savefig(\n",
    "    f\"figures/show/somde/{DATA_FOLDER_NAME}/ini_codebook.png\",\n",
    "    dpi=300,\n",
    "    bbox_inches=\"tight\"\n",
    ")"
   ],
   "id": "34a9c0c9f9e4a07e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Plot the Trained SOM\n",
    "\n",
    "This cell visualizes the **trained Self-Organizing Map (SOM)** and the mapping of data points onto it."
   ],
   "id": "8c4375f480319569"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Plot the SOM (creates its own figure)\n",
    "som.view()\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(5, 5)\n",
    "fig.savefig(\n",
    "    f\"figures/show/somde/{DATA_FOLDER_NAME}/somde_points.png\",\n",
    "    dpi=300,\n",
    "    bbox_inches=\"tight\"\n",
    ")"
   ],
   "id": "8f222c94f5b5e49e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Map Expression Data onto the SOM\n",
    "\n",
    "This cell projects the **gene expression matrix** onto the trained SOM."
   ],
   "id": "da65dbfd138d71d7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "ndf, ninfo = som.mtx(df.T)\n",
    "print(ndf.shape)\n",
    "print(ninfo.shape)\n",
    "\n",
    "genes = df.columns\n"
   ],
   "id": "6e6dcdb10b7f4577",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Visualize Gene Expression on the SOM\n",
    "\n",
    "This cell plots the spatial distribution of selected genes on the SOM."
   ],
   "id": "81af1c0354802116"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "plotgene(\n",
    "    X=ninfo.values,\n",
    "    mtx=ndf,\n",
    "    draw_list=genes[:5],\n",
    "    result=som,\n",
    "    sp=100,\n",
    "    lw=0.2,\n",
    "    N=10,\n",
    "    plotsize=5\n",
    ")"
   ],
   "id": "5e5f5314df6e0e8f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Normalize and Run SOMDE Analysis\n",
    "\n",
    "This cell performs **normalization** and runs the **SOMDE algorithm** on the dataset."
   ],
   "id": "d1fb05a4169aeda5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "nres = som.norm()\n",
    "\n",
    "print(\"SOMDE analysis\")\n",
    "\n",
    "process = psutil.Process(os.getpid())\n",
    "process.cpu_percent(interval=None)   # reset counters\n",
    "\n",
    "start_cpu = time.process_time()\n",
    "start_wall = time.time()\n",
    "\n",
    "# -------- code to measure -----------\n",
    "result, SVnum =som.run()\n",
    "# ------------------------------------\n",
    "\n",
    "end_cpu = time.process_time()\n",
    "end_wall = time.time()\n",
    "\n",
    "cpu_time = end_cpu - start_cpu\n",
    "wall_time = end_wall - start_wall\n",
    "cpu_usage = process.cpu_percent(interval=None)\n",
    "ram_usage = process.memory_info().rss / (1024**2)  # MB\n",
    "\n",
    "print(\"CPU time: {:.6f} seconds\".format(cpu_time))\n",
    "print(\"Wall time: {:.6f} seconds\".format(wall_time))\n",
    "print(\"CPU usage (%):\", cpu_usage)\n",
    "print(\"RAM used by process: {:.2f} MB\".format(ram_usage))\n",
    "\n",
    "# ---------------- SAVE TO CSV -----------------\n",
    "with open(os.path.join(output_dir,  \"requirements.csv\"), \"a\", newline=\"\") as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow([\"somde analysis\", cpu_time, wall_time, cpu_usage, ram_usage])"
   ],
   "id": "ec06c619155c0962",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Visualize Spatially Variable Genes from SOMDE\n",
    "\n",
    "This cell plots the expression patterns of selected **spatially variable genes (SVGs)**"
   ],
   "id": "3ccef26be6278d8f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "genes_to_display = result.g.to_list()[:5]\n",
    "\n",
    "plotgene(ninfo[['x','y']].values,nres.T,genes_to_display,result,sp=80,lw=0.2,plotsize=5)"
   ],
   "id": "623692f2171b1f05",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Plot Selected Spatially Variable Genes on Original Coordinates\n",
    "\n",
    "This cell visualizes the expression patterns of a subset of **spatially variable genes (SVGs)** on the **original spatial coordinates**."
   ],
   "id": "9db7dd48e790bd03"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "genes_to_display = result.g.to_list()[:5]\n",
    "plotgene(corinfo[['x','y']].values,df.T,genes_to_display,result,sp=2,lw=0.2,plotsize=5)"
   ],
   "id": "a0edac4f0a694813",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Display Number of Spatially Variable Genes",
   "id": "3be509169778aa17"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Display SOMDE Results Object\n",
    "\n",
    "This cell outputs the **full SOMDE results object**."
   ],
   "id": "614d3e96d88f5d30"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "result",
   "id": "78b20fdaea0c2b1d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "result.to_csv(os.path.join(output_dir,  \"somde_result.csv\"), index=True)",
   "id": "875b4f417ba0cf83",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Spatiallt variable genes\n",
    "qval <= 0.05 and FSV >= 0.995"
   ],
   "id": "c5ac95ea202f07f4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "SV_genes = result[(result['qval'] < 0.05) & (result['FSV'] >= 0.995)]\n",
    "SV_genes.to_csv(os.path.join(output_dir, \"somde_result.SV_genes.qval+FSV.csv\"), index=True)\n",
    "print(SV_genes)\n",
    "print(len(SV_genes[\"g\"].to_list()))"
   ],
   "id": "b11b7ad2e7af937c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "SV_genes_somde = result[(result['qval'] < 0.05)]\n",
    "SV_genes_somde.to_csv(os.path.join(output_dir, \"somde_result.SV_genes.qval.csv\"), index=True)\n",
    "\n",
    "print(len(SV_genes_somde[\"g\"].to_list()), SVnum)"
   ],
   "id": "d9ed3b5ac41a7dc0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "print(SVnum)\n",
    "\n",
    "with open(os.path.join(output_dir,\"SVnum.csv\"), \"w\", newline=\"\") as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow([\"SVnum_somde\", \"SVnum\"])\n",
    "    writer.writerow([SVnum, len(SV_genes[\"g\"].to_list())])"
   ],
   "id": "df544d22afbeb6c4",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (somde_x86)",
   "language": "python",
   "name": "somde_x86"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
